### DISCUSSION TITLE
Introducing concepts

### DISCUSSION THREAD

**Discussion by ftomassetti** [Strumenta/StarLasu#20] [General]
At the moment nodes know their type and their list of properties. We should introduce proper concepts also and nodes should be able to return their concepts.

**Comment by alessiostalla** [comment]
Tylasu has [NodeDefinition](https://github.com/Strumenta/tylasu/blob/6b1285192d29b322e40b346cd2f36f0c5784b76d/src/model/model.ts#L14), it's an internal API.

**Reply by ftomassetti** [reply]
Good! For version 2 we could extend it, possibly by also using a certain project that could come up around the end of the year with some code which is also publicly visible here: https://github.com/LIonWeb-org/lioncore-typescript/tree/feature/m3

**Reply by ftomassetti** [reply]
More specifically this: https://github.com/LIonWeb-org/lioncore-typescript/blob/e51a209ef51d5cfbb42857896c5e38ec59611047/src/m3/types.ts#L36

**Comment by alessiostalla** [comment]
Some thoughts.

A Concept would be a kind of "metaclass", should it also be a "meta Node" i.e. be an instance of some Node subclass that happens to describe another Node? 

Concepts should be grouped in some kind of container/package/namespace/scope thing. That makes it more palatable to have Concepts as Nodes because also their container would be a Node and we could use regular *lasu APIs to search for nodes/concepts.

**Comment by ftomassetti** [comment]
> A Concept would be a kind of "metaclass", should it also be a "meta Node" i.e. be an instance of some Node subclass that happens to describe another Node?

I think this would be related to the idea of "meta-circularity". In general yes, we may want to be able to provide a representation of concepts as nodes, so that we can serialize them in the same way that we serialize nodes. I am not sure if the Concept class itself should extend Node or if we could export Concepts to Nodes.

> Concepts should be grouped in some kind of container/package/namespace/scope thing.

Yes. We could call it also "language" or "metamodel". But it would be effectively a namespace

> That makes it more palatable to have Concepts as Nodes because also their container would be a Node and we could use regular *lasu APIs to search for nodes/concepts.

Yes, indeed. So metamodels would be models that happen to contains nodes of a particular type (concepts)
